### DISCUSSION TITLE
The id must be an integer che mi perseguita

### DISCUSSION THREAD

**Discussion by francescoc100** [fattureincloud/api#134] [Q&A] [answered]
Ciao, uso le API tramite token, nella creazione di una fattura, usando il codice di esempio e personalizzando trovo sempre l'errore: int(422) string(98) "{"error":{"message":"Invalid request.","validation_result":{"id":["The id must be an integer."]}}}" 

Peraltro, altri errori simili nelle API reference del sito ufficiale...riporto l'array che preparo

$data = array('data'=> array('type'=> 'invoice','entity'=> array('id'=> '1','name'=> 'Mario Rossi','vat_number'=> '47803200154','tax_code'=> 'RSSMRA91M20B967Q','address_street'=> 'Via Italia, 66','address_postal_code'=> '20900','address_city'=> 'Milano','address_province'=> 'MI','address_extra'=> '','country'=> 'Italia'),'date'=> '2022-01-20','number'=> '1','numeration'=> '/01','subject'=> 'internal subject','visible_subject'=> 'visible subject','currency'=> array('id'=> 'EUR','exchange_rate'=> '1.00000','symbol'=> '€'),'language'=> array('code'=> 'it','name'=> 'Italiano'),'items_list'=> [array('product_id'=> '4','code'=> 'TV3','name'=> 'Tavolo in legno','net_price'=> '100','category'=> 'cucina','discount'=> '0','qty'=> '1','vat'=> array('id'=> '0'))],'payment_method'=>array('id'=> 'MP05'),'payments_list'=> [array('amount'=> '122','due_date'=> '2022-01-23','paid_date'=> '2022-01-22','status'=> 'paid','payment_account'=> array('id'=> '1'))],'attachment_token'=> 'YmMyNWYxYzIwMTU3N2Y4ZGE3ZjZiMzg5OWY0ODNkZDQveXl5LmRvYw','template'=> array('id'=> '150')));

**Comment by sensorario** [comment, marked as answer]
Vedo che tutti gli id li passi come stringa. Ad esempio `array('id' => '1'`. Potrebbe essere questo il problema?

**Reply by valmoz** [reply]
Ciao @francescoc100 
come ha detto @sensorario (a proposito, grazie!) penso anche io che il problema sia quello.

Quando possibile, noi suggeriamo sempre di utilizzare le [nostre SDK](https://developers.fattureincloud.it/docs/sdks/php-sdk), che dovrebbero guidarvi nella compilazione della richiesta API e limitare questo genere di problemi.

Suggerisco anche la [guida alla creazione di una fattura](https://developers.fattureincloud.it/docs/guides/invoice-creation), che è uno dei nostri bestseller :)

**Reply by francescoc100** [reply]
Grazie mille! Si ho seguito la guida, pagina indispensabile :) Tuttavia non sono molto pratico nell'uso delle SDK...e ci sto provando a mano.
Vi chiedo gentilmente, se possibile, di spiegarmi con qualche dato in più l'errore che avete trovato, vi riposto il codice con qualche correzzione presa dal vostro esempio, di cui cmq compare lo stesso errore.

$data = array('data'=> array('type'=> 'invoice','entity'=> array('name'=> 'Mario Rossi','vat_number'=> '47803200154','tax_code'=> 'RSSMRA91M20B967Q','address_street'=> 'Via Italia, 66','address_postal_code'=> '20900','address_city'=> 'Milano','address_province'=> 'MI','address_extra'=> '','country'=> 'Italia'),'date'=> '2022-01-20','numeration'=> '/01','subject'=> 'internal subject','visible_subject'=> 'visible subject','currency'=> array('id'=> 'EUR','exchange_rate'=> '1.00000','symbol'=> '€'),'language'=> array('code'=> 'it','name'=> 'Italiano'),'items_list'=> [array('code'=> 'TV3','name'=> 'Tavolo in legno','net_price'=> '100','category'=> 'cucina','discount'=> '0','qty'=> '1','vat'=> array('id'=> '0'))],'payment_method'=>array('id'=> 'MP05'),'payments_list'=> [array('amount'=> '122','due_date'=> '2022-01-23','paid_date'=> '2022-01-22','status'=> 'paid','payment_account'=> array('id'=> '1'))]));

Per dire, su altre pagine come prodotti o clienti sono riuscito.
Grazie per l'aiuto!!!!

**Reply by emajo** [reply]
Ciao @francescoc100 dal codice che hai passato si vedono ancora molti campi che sono numerici passati invece come stringa, per farti qualche esempio `net_price`, `qty`, `discount`, `vat id`, `amount`, `payment_account id`.

Per costruire la richiesta ti consiglio di fare più attenzione ai tipi dei campi controllando lo schema di ogni richiesta sulla nostra [API Reference](https://developers.fattureincloud.it/api-reference#post-/c/-company_id-/issued_documents), oppure potresti utilizzare l'SDK che va a mitigare in partenza questi problemi.

Grazie,
Emanuele.

**Reply by valmoz** [reply]
Buongiorno @francescoc100 sei riuscito a risolvere?
